Public/Src/Tools/SymbolDaemon/VsoSymbolClient.cs

```diff
@@ -0,0 +1,208 @@
+{
+    /// <summary>
+    /// Client for Artifact Services symbols endpoint.
+    /// </summary>
+    public sealed class VsoSymbolClient : ISymbolClient
+    {
+        private const DebugEntryCreateBehavior DefaultDebugEntryCreateBehavior = DebugEntryCreateBehavior.ThrowIfExists;
+
+        private static IAppTraceSource Tracer => SymbolAppTraceSource.SingleInstance;
+
+        private readonly ILogger m_logger;
+        private readonly SymbolConfig m_config;
+        private readonly ISymbolServiceClient m_symbolClient;
+        private readonly CancellationTokenSource m_cancellationSource;
+        private CancellationToken Token => m_cancellationSource.Token;
+        private string m_requestId;
+
+        private VssCredentials GetCredentials() =>
+            new VsoCredentialHelper(m => m_logger.Verbose(m))
+                .GetCredentials(m_config.Service, true, null);
+
+        private ArtifactHttpClientFactory GetFactory() =>
+            new ArtifactHttpClientFactory(
+                credentials: GetCredentials(),
+                httpSendTimeout: m_config.HttpSendTimeout,
+                tracer: Tracer,
+                verifyConnectionCancellationToken: Token);
+
+        private Uri ServiceEndpoint => m_config.Service;
+
+        private string RequestName => m_config.Name;
+
+        private string RequestId
+        {
+            get
+            {
+                Contract.Requires(!string.IsNullOrEmpty(m_requestId));
+                return m_requestId;
+            }
+        }
+
+        /// <nodoc />
+        public VsoSymbolClient(ILogger logger, SymbolConfig config)
+        {
+            m_logger = logger;
+            m_config = config;
+            m_cancellationSource = new CancellationTokenSource();
+
+            m_logger.Info(I($"[{nameof(VsoSymbolClient)}] Using drop config: {JsonConvert.SerializeObject(m_config)}"));
+
+            m_symbolClient = CreateSymbolServiceClient();
+        }
+
+        private ISymbolServiceClient CreateSymbolServiceClient()
+        {
+            var client = new SymbolServiceClient(
+                ServiceEndpoint,
+                GetFactory(),
+                Tracer,
+                new SymbolServiceClientTelemetry(Tracer, ServiceEndpoint, enable: m_config.EnableTelemetry));
+
+            return client;
+        }
+
+        /// <summary>
+        /// Queries the endpoint for the RequestId. 
+        /// This method should be called only after the request has been created, otherwise, it will throw an exception.
+        /// </summary>
+        /// <remarks>
+        /// On workers, m_requestId won't be initialized, so we need to query the server for the right value.
+        /// </remarks>
+        private async Task EnsureRequestIdInitalizedAsync()
+        {
+            if (string.IsNullOrEmpty(m_requestId))
+            {
+                var result = await m_symbolClient.GetRequestByNameAsync(RequestName, Token);
+                m_requestId = result.Id;

```

Comment:
> m_requestId = result.Id; [](start = 15, length = 25)

Why do we re-assign `m_requestId`? Should we not just compare with what `CreateAsync` produced?

Reply:
because `CreateAsync` might have run on a different worker. So if we need to execute a symbol command on a worker, and that worker does not know yet the value for `m_requestId`, it will politely ask the endpoint about this.

---
In reply to: [313641961](https://github.com/microsoft/BuildXL/pull/727#discussion_r313641961) [](ancestors = 313641961)